src/piton/jax.py

```diff
@@ -0,0 +1,1048 @@
+"""
+A collection of custom jax primitives that are necessary for piton's neural network models to operate.
+"""
+
+from __future__ import annotations
+
+import ctypes
+import logging
+import struct
+import warnings
+from functools import partial
+from typing import Any, Dict, Optional, Sequence, Tuple, cast
+
+import jax
+import numpy as np
+from jax import core, custom_vjp, debug, grad, lax, nn
+from jax import numpy as jnp
+from jax import value_and_grad, vmap, xla_computation
+from jax.core import ConcreteArray, ShapedArray
+from jax.interpreters import ad, batching, xla
+from jax.lib import xla_client
+
+# Per the jax documentation, we currently don't have good typing for arrays
+Array = Any
+
+from piton.extension.jax import (
+    get_kernels,
+    get_local_attention_data,
+    get_local_attention_shape,
+)
+
+# Globally register all of our custom operators
+for name, value, platform in get_kernels():
+    jax.lib.xla_client.register_custom_call_target(
+        name, value, platform=platform
+    )
+
+
+@partial(
+    jax.custom_vjp,
+    nondiff_argnums=(2,),
+)
+def gather_scatter(a: Array, indices: Array, output_dim: int) -> Array:
+    """Perform a combined gather / scatter add operation.
+
+    Indices is an array of shape (N, 2) for N indices.
+    This operator reads from the index specified in indices[:, 0] and writes them to indices[:, 1].
+
+    output_dim is the size of the ouptupt array.
+
+    Note that this operator requires that indices must be sorted in the order of the write indices (indices[:, 1])
+    """
+    return gather_scatter_fwd(a, indices, output_dim)[0]
+
+
+def gather_scatter_fallback(a: Array, indices: Array, output_dim: int) -> Array:
+    """The python fallback implementation of gather scatter."""
+    a_indices = indices[:, 0]
+    b_indices = indices[:, 1]
+
+    a_vals = a.at[a_indices, :].get(mode="fill", fill_value=0)
+
+    result = jnp.zeros((output_dim, a.shape[1]), dtype=a.dtype)
+
+    return result.at[b_indices, :].add(a_vals, mode="drop")
```

Comment:
Comment: 

```
The `.add()` here successively adds values in `a_vals` that map to the same index (i.e. have the same associated value in `b_indices`) in a non-deterministic order (which is OK since addition is commutative).

Concrete example:
b_indices = [ 0, 1, 1, 1 ]
a_vals = [ 1, 1, 1, 1, ]
result = [ 1, 3 ]
```

Reply:
Added to the main documentation.